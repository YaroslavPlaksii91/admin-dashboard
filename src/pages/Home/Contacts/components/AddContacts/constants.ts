import { IMAGE_SIZE } from '@components/AddPhoto/constants';

export const ADD_CONTACT_FIELDS = Object.freeze({
  PHOTO: 'photo',
  FIRST_NAME: 'firstName',
  LAST_NAME: 'lastName',
  EMAIL: 'email',
  ADDRESS: 'address',
});

export const ADD_CONTACT_CONFIG = Object.freeze({
  [ADD_CONTACT_FIELDS.PHOTO]: {
    label: 'Add photo',
    name: 'photo',
    size: IMAGE_SIZE.BIG,
  },
  [ADD_CONTACT_FIELDS.FIRST_NAME]: {
    label: 'first name',
    type: 'text',
    name: 'firstName',
    placeholder: 'First name',
  },
  [ADD_CONTACT_FIELDS.LAST_NAME]: {
    label: 'last name',
    type: 'text',
    name: 'lastName',
    placeholder: 'Last name',
  },
  [ADD_CONTACT_FIELDS.EMAIL]: {
    label: 'email',
    type: 'email',
    name: 'email',
    placeholder: 'Email',
  },
  [ADD_CONTACT_FIELDS.ADDRESS]: {
    label: 'address',
    type: 'text',
    name: 'address',
    placeholder: 'Address',
  },
});
