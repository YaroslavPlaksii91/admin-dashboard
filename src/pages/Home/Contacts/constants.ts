export const CONTACTS_COLUMNS = Object.freeze([
  {
    title: 'Name',
    name: 'name',
    size: 4,
  },
  {
    title: 'Email',
    name: 'email',
    size: 3,
  },
  {
    title: 'Address',
    name: 'address',
    size: 3,
  },
  {
    title: 'Created at',
    name: 'created',
    size: 2,
  },
]);
