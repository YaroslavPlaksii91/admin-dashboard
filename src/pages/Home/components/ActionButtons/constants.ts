export const SORT_BTN_TEST_ID = 'sortButton';
export const FILTER_BTN_TEST_ID = 'filterButton';
export const ADD_BTN_TEST_ID = 'addButton';
