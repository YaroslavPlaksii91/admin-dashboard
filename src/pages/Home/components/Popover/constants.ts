export const POPOVER_TEST_ID = 'popover';
