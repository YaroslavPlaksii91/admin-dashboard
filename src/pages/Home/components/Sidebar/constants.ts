export const USERS_PAGES = Object.freeze([
  {
    name: 'Overview',
    icon: '/sprite.svg#icon-overview',
    link: '/overview',
  },
  {
    name: 'Tickets',
    icon: '/sprite.svg#icon-tickets',
    link: '/tickets',
  },
  {
    name: 'Ideas',
    icon: '/sprite.svg#icon-ideas',
    link: '/ideas',
  },
  {
    name: 'Contacts',
    icon: '/sprite.svg#icon-contacts',
    link: '/contacts',
  },
  {
    name: 'Agents',
    icon: '/sprite.svg#icon-agents',
    link: '/agents',
  },
  {
    name: 'Articles',
    icon: '/sprite.svg#icon-articles',
    link: '/articles',
  },
]);

export const USERS_SETTINGS = Object.freeze([
  {
    name: 'Settings',
    icon: '/sprite.svg#icon-settings',
    link: '/settings',
  },
  {
    name: 'Subscription',
    icon: '/sprite.svg#icon-subscription',
    link: '/subscription',
  },
]);

export const SIDEBAR_LINK_TEST_ID = 'sidebarLink';
