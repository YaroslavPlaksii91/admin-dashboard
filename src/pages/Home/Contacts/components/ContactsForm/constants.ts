import { IMAGE_SIZE } from '@components/AddPhoto/constants';

export const CONTACTS_FORM_FIELDS = Object.freeze({
  PHOTO: 'photo',
  FIRST_NAME: 'firstName',
  LAST_NAME: 'lastName',
  EMAIL: 'email',
  ADDRESS: 'address',
});

export const CONTACTS_FORM_CONFIG = Object.freeze({
  [CONTACTS_FORM_FIELDS.PHOTO]: {
    label: 'Add photo',
    name: 'photo',
    size: IMAGE_SIZE.BIG,
  },
  [CONTACTS_FORM_FIELDS.FIRST_NAME]: {
    label: 'first name',
    type: 'text',
    name: 'firstName',
    placeholder: 'First name',
  },
  [CONTACTS_FORM_FIELDS.LAST_NAME]: {
    label: 'last name',
    type: 'text',
    name: 'lastName',
    placeholder: 'Last name',
  },
  [CONTACTS_FORM_FIELDS.EMAIL]: {
    label: 'email',
    type: 'email',
    name: 'email',
    placeholder: 'Email',
  },
  [CONTACTS_FORM_FIELDS.ADDRESS]: {
    label: 'address',
    type: 'text',
    name: 'address',
    placeholder: 'Address',
  },
});
