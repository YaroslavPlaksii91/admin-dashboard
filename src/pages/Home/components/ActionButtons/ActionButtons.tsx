import { FC, useState, MouseEvent } from 'react';
import { Box, Button, SvgIcon } from '@mui/material';

import { ActionButtonsProps } from './types';
import {
  SORT_BTN_TEST_ID,
  FILTER_BTN_TEST_ID,
  ADD_BTN_TEST_ID,
} from './constants';
import { Popover } from '../Popover/Popover';
import { RadioButtons } from '../RadioButtons/RadioButtons';

export const ActionButtons: FC<ActionButtonsProps> = ({
  handleSort,
  onAddClick,
  addButtonName,
  sortOptions,
  setFilterValue,
  filterValue,
  filterTitle,
  filterOptions,
}) => {
  const [sortAnchorEl, setSortAnchorEl] = useState<HTMLButtonElement | null>(
    null,
  );
  const [filterAnchorEl, setFilterAnchorEl] =
    useState<HTMLButtonElement | null>(null);

  const handleSortClick = (event: MouseEvent<HTMLButtonElement>) => {
    setSortAnchorEl(event.currentTarget);
  };
  const handleFilterClick = (event: MouseEvent<HTMLButtonElement>) => {
    setFilterAnchorEl(event.currentTarget);
  };

  const handleSortClose = () => {
    setSortAnchorEl(null);
  };
  const handleFilterClose = () => {
    setFilterAnchorEl(null);
  };

  const sortOpen = Boolean(sortAnchorEl);
  const sortId = sortOpen ? 'sort-popover' : undefined;

  const filterOpen = Boolean(filterAnchorEl);
  const filterId = filterOpen ? 'filter-popover' : undefined;

  const handleFilterChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFilterValue((event.target as HTMLInputElement).value);
  };

  return (
    <Box sx={{ display: 'flex', mb: '47px', padding: '0 32px' }}>
      <Button
        variant="text"
        sx={{ color: 'grayDarkColor', mr: 8 }}
        aria-describedby={sortId}
        onClick={handleSortClick}
        data-testid={SORT_BTN_TEST_ID}
      >
        <SvgIcon titleAccess="Sort" fontSize="small" sx={{ mr: 2 }}>
          <use href="/sprite.svg#icon-sort"></use>
        </SvgIcon>
        Sort
      </Button>

      <Popover
        id={sortId}
        open={sortOpen}
        anchorEl={sortAnchorEl}
        handleClose={handleSortClose}
      >
        {sortOptions.map(option => (
          <Button
            key={option}
            variant="text"
            sx={{ display: 'block', width: '100%', padding: '10px' }}
            onClick={() => handleSort(option)}
          >
            By {option}
          </Button>
        ))}
      </Popover>

      <Button
        variant="text"
        sx={{ color: 'grayDarkColor' }}
        aria-describedby={filterId}
        onClick={handleFilterClick}
        data-testid={FILTER_BTN_TEST_ID}
      >
        <SvgIcon titleAccess="Filter" fontSize="small" sx={{ mr: 2 }}>
          <use href="/sprite.svg#icon-filter"></use>
        </SvgIcon>
        Filter
      </Button>

      <Popover
        id={filterId}
        open={filterOpen}
        anchorEl={filterAnchorEl}
        handleClose={handleFilterClose}
      >
        <RadioButtons
          value={filterValue}
          handleChange={handleFilterChange}
          title={filterTitle}
          options={filterOptions}
        />
      </Popover>

      <Button
        variant="text"
        sx={{ color: 'accentBlueColor', ml: 'auto' }}
        onClick={onAddClick}
        data-testid={ADD_BTN_TEST_ID}
      >
        <SvgIcon titleAccess="Add new item" sx={{ mr: 2, width: 8, height: 8 }}>
          <use href="/sprite.svg#icon-add"></use>
        </SvgIcon>
        {addButtonName}
      </Button>
    </Box>
  );
};
