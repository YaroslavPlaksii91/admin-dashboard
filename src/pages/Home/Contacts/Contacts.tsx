import { FC, useState, useEffect } from 'react';
import { Box } from '@mui/material';

import { Pagination } from '@components/Pagination/Pagination';
import { usePagination } from '@components/Pagination/usePagination';
import { ModalComponent } from '@components/Modal/Modal';
import { getContacts } from '@services/db/getContacts';
import { useSort } from '@hooks/useSort';

import { ContactType } from './components/ContactsItem/types';
import { ContactsItem } from './components/ContactsItem/ContactsItem';
import { AddContacts } from './components/AddContacts/AddContacts';
import { AddContactsData } from './components/AddContacts/types';
import { currentDate } from './helpers/getCurrentDate';
import { CONTACTS_COLUMNS } from './constants';
import { Heading } from '../components/Heading/Heading';
import { ActionButtons } from '../components/ActionButtons/ActionButtons';

export const Contacts: FC = () => {
  const [contacts, setContacts] = useState<ContactType[]>([]);
  const [isModalOpen, setIsModalOpen] = useState(false);

  const { sortedData, handleSort, sortKey, sortDirection } = useSort(contacts);

  const { getCurrentPageData, page, setPage, rowsPerPage, setRowsPerPage } =
    usePagination(sortedData);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const data = await getContacts();
        setContacts(data);
      } catch (error) {
        console.error('Error while fetching data: ', error);
      }
    };

    fetchData();
  }, []);

  const handleAddContact = ({
    firstName,
    lastName,
    email,
    address,
    photo,
  }: AddContactsData) => {
    const newContact = {
      id: contacts.length + 1,
      name: `${firstName} ${lastName}`,
      email,
      address,
      created: currentDate,
      image: URL.createObjectURL(photo[0]),
    };

    setContacts([newContact, ...contacts]);

    setIsModalOpen(false);
  };

  const onSortClick = () => {};

  const onFilterClick = () => {
    console.log('Filter');
  };

  return (
    <Box
      sx={{
        height: '100%',
        pt: '37px',
        border: '1px solid',
        borderColor: 'borderColor',
        borderRadius: '8px',
        backgroundColor: 'mainWhiteColor',
      }}
    >
      <ActionButtons
        addButtonName="Add contact"
        onAddClick={() => setIsModalOpen(true)}
        onFilterClick={onFilterClick}
        onSortClick={onSortClick}
      />

      <Heading
        handleSort={handleSort}
        sortKey={sortKey}
        sortDirection={sortDirection}
        columns={CONTACTS_COLUMNS}
      />

      {getCurrentPageData().map(contact => (
        <ContactsItem contact={contact} key={contact.id} />
      ))}

      <Pagination
        count={contacts.length}
        page={page}
        setPage={setPage}
        rowsPerPage={rowsPerPage}
        setRowsPerPage={setRowsPerPage}
      />

      <ModalComponent
        isOpen={isModalOpen}
        title="Add new contact"
        onClose={() => setIsModalOpen(false)}
      >
        <AddContacts addContact={handleAddContact} />
      </ModalComponent>
    </Box>
  );
};
