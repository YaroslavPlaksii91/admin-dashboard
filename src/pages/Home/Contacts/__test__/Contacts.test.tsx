import { render, cleanup, fireEvent, screen } from '@testing-library/react';

import { ADD_BTN_TEST_ID } from '@pages/Home/components/ActionButtons/constants';

import { Contacts } from '../Contacts';

jest.mock('@services/db/contacts', () => ({
  getContacts: jest.fn(() => Promise.resolve([])),
  createContact: jest.fn(newContact => Promise.resolve(newContact)),
  deleteContact: jest.fn(contactId => Promise.resolve(contactId)),
  editContact: jest.fn((_, updatedData) => Promise.resolve(updatedData)),
}));

jest.mock('@services/date/formatDate', () => ({
  formatDate: jest.fn(date => ({ date })),
}));

describe('Contacts component', () => {
  afterEach(cleanup);

  it('should be rendered', () => {
    const { asFragment } = render(<Contacts />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('opens the modal when "Add contact" button is clicked', () => {
    render(<Contacts />);

    fireEvent.click(screen.getByTestId(ADD_BTN_TEST_ID));

    expect(screen.getByText('Add new contact')).toBeInTheDocument();
  });
});
