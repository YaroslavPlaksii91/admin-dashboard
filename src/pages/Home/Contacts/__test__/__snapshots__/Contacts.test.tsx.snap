// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts component should be rendered 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-zqkb86"
  >
    <div
      class="MuiBox-root css-la04gv"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-17u4eyv-MuiButtonBase-root-MuiButton-root"
        data-testid="sortButton"
        tabindex="0"
        type="button"
      >
        <svg
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1v1n4yu-MuiSvgIcon-root"
          focusable="false"
          role="img"
          viewBox="0 0 24 24"
        >
          <use
            href="/sprite.svg#icon-sort"
          />
          <title>
            Sort
          </title>
        </svg>
        Sort
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-dpl43n-MuiButtonBase-root-MuiButton-root"
        data-testid="filterButton"
        tabindex="0"
        type="button"
      >
        <svg
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1v1n4yu-MuiSvgIcon-root"
          focusable="false"
          role="img"
          viewBox="0 0 24 24"
        >
          <use
            href="/sprite.svg#icon-filter"
          />
          <title>
            Filter
          </title>
        </svg>
        Filter
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-e9wzt7-MuiButtonBase-root-MuiButton-root"
        data-testid="addButton"
        tabindex="0"
        type="button"
      >
        <svg
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1vh51qy-MuiSvgIcon-root"
          focusable="false"
          role="img"
          viewBox="0 0 24 24"
        >
          <use
            href="/sprite.svg#icon-add"
          />
          <title>
            Add new item
          </title>
        </svg>
        Add contact
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-bxay1f-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
      >
        Name 
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
      >
        Email 
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1equabv-MuiGrid-root"
      >
        Address 
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1o7apob-MuiGrid-root"
      >
        Created at 
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1j6x8sl-MuiSvgIcon-root"
          data-testid="SouthIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m19 15-1.41-1.41L13 18.17V2h-2v16.17l-4.59-4.59L5 15l7 7 7-7z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiTablePagination-root css-1rfvjt6-MuiTablePagination-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
      >
        <div
          class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
        />
        <p
          class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
          id=":r1:"
        >
          Rows per page:
        </p>
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-16c50h-MuiInputBase-root-MuiTablePagination-select"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby=":r1: :r0:"
            class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
            id=":r0:"
            role="button"
            tabindex="0"
          >
            8
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="8"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
        >
          0â€“0 of 0
        </p>
        <div
          class="MuiTablePagination-actions"
        >
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to previous page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowLeftIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </button>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            disabled=""
            tabindex="-1"
            title="Go to next page"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
